packages/database/src/migrations/20201009061046-UpdateLaosEocEntities-modifies-data.js

```diff
@@ -0,0 +1,257 @@
+      lon: 102.133333,
+    },
+  },
+  Oudomxay: {
+    capital: {
+      name: 'Xai',
+      lat: 20.691389,
+      lon: 101.986111,
+    },
+  },
+  Phongsaly: {
+    capital: {
+      name: 'Phongsali',
+      lat: 21.683333,
+      lon: 102.1,
+    },
+  },
+  Xainyabuli: {
+    capital: {
+      name: 'Xainyabuli',
+      lat: 19.416667,
+      lon: 103.166667,
+    },
+  },
+  Salavan: {
+    capital: {
+      name: 'Salavan',
+      lat: 15.716667,
+      lon: 106.416667,
+    },
+  },
+  Savannakhet: {
+    capital: {
+      name: 'Kaysone Phomvihane',
+      lat: 16.55,
+      lon: 104.75,
+    },
+  },
+  Xekong: {
+    capital: {
+      name: 'Lamarn',
+      lat: 15.344444,
+      lon: 106.7175,
+    },
+  },
+  'Vientiane Capital': {
+    capital: {
+      name: 'Sisattanak',
+      lat: 17.93219235318203,
+      lon: 102.63719558715822,
+    },
+  },
+  'Vientiane Province': {
+    capital: {
+      name: 'Phônhông',
+      lat: 18.5,
+      lon: 102.416667,
+    },
+  },
+  Xiangkhouang: {
+    capital: {
+      name: 'Pek',
+      lat: 19.416667,
+      lon: 103.166667,
+    },
+  },
+  Xaisomboun: {
+    capital: {
+      name: 'Anouvong',
+      lat: 18.9057,
+      lon: 103.092,
+    },
+  },
+};
+
+// Pre-calculate entity codes (needed for both up() and down()), and id
+for (const provinceName of Object.keys(PROVINCES)) {
+  const { capital } = PROVINCES[provinceName];
+  const uppercaseName = capital.name.toUpperCase().replace(' ', '_');
+  PROVINCES[provinceName].capital.code = `LA_PROV_CAPITAL_${uppercaseName}`;
+  PROVINCES[provinceName].capital.id = generateId();
+}
+
+const getLaosCountryEntityId = async function (db) {
+  const results = await db.runSql(`SELECT id FROM entity WHERE code = 'LA';`);
+
+  if (results.rows.length > 0) {
+    return results.rows[0].id;
+  }
+
+  throw new Error('Entity not found');
+};
+
+const getEntityHierarchyId = async function (db) {
+  const results = await db.runSql(`SELECT id FROM entity_hierarchy WHERE name = 'laos_eoc';`);
+
+  if (results.rows.length > 0) {
+    return results.rows[0].id;
+  }
+
+  throw new Error('Entity hierarchy not found');
+};
+
+const clearData = async function (db) {
+  await db.runSql(
+    "DELETE FROM entity_relation where child_id IN (SELECT id FROM entity WHERE code like 'LA_PROV_CAPITAL_%');",
+  );
+  await db.runSql("DELETE FROM entity WHERE code like 'LA_PROV_CAPITAL_%';");
+};
+
+exports.up = async function (db) {
+  const parentId = await getLaosCountryEntityId(db);
+
+  const entityHierarchyId = await getEntityHierarchyId(db);
+
+  // Delete old data
+  await clearData(db);
+
+  /*
+   * Entity
+   */
+  let sql =
+    'INSERT INTO entity (id, code, parent_id, name, type, point, country_code, bounds, metadata) VALUES \n';
+
+  const rowsSql = [];
+
+  for (const province of Object.values(PROVINCES)) {
+    let rowSql = '(';
+
+    const point = { type: 'Point', coordinates: [province.capital.lon, province.capital.lat] };
+
+    rowSql +=
```

Comment:
Is performance such an issue/priority here? Don't get me wrong, I love building the sql like this!

However, given that it is just a migration perhaps it will be more readable as a single `db.runSql` inside the loop, similar to how you've done `entity_relation`? (I assume that you can't insert geoJson with `insertObject`, otherwise that would be ideal).

If it's too messy, you could extract it to a `insertEntity` function.

Reply:
Ah yes I would love to use insertEntity but it doesn't play nice with inserting geojson fields :(